File: Data Structures & Algorithms/sliding-window-maximum/submission-2.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        l = 0
        window = []
        output = []

        for r in range(len(nums)):
            if r - l == k:
                window.remove(nums[l])
                l += 1
            
            window.append(nums[r])

            if r - l + 1 == k:
                output.append(max(window))
        
        return output

File: Data Structures & Algorithms/maximum-product-subarray/submission-2.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int n = nums.size();
        if (n == 1) return nums[0];

        vector<vector<long int>> dp(n, vector<long int>(n, 1));

        for (int i = 0; i < n; ++i) {
            dp[i][i] = nums[i];
        }

        long int maxProduct = INT_MIN;
        for (int len = 2; len <= n; ++len) {
            for (int i = 0; i + len <= n; ++i) {
                int j = i + len - 1;

                dp[i][j] = dp[i][j-1] * nums[j];

                maxProduct = max(maxProduct, dp[i][j]);
            }
        }

        return maxProduct;
    }
};
